### ISSUE TITLE
Warp sync not respecting warp-barrier

### ISSUE THREAD

**Issue by maxsam4** [paritytech/parity-ethereum#9901]
- **Parity Ethereum version**: v2.0.6-unstable-549e202-20180919 and v2.1.4-beta-bee2cb8-20181028
- **Operating system**: Linux 18.04
- **Installation**: snap store / one line installer
- **Fully synchronized**:  no
- **Network**: ethereum
- **Restarted**: yes

**Actual behavior**: Parity ignores the `warp-barrier` option
```
max@max-GF63-8RC:~/snap/parity/15594/.local/share/io.parity.ethereum$ parity --warp-barrier=100000 --no-ancient-blocks
2018-11-13 14:17:42  Starting Parity-Ethereum/v2.0.6-unstable-549e202-20180919/x86_64-linux-gnu/rustc1.29.0
2018-11-13 14:17:42  Keys path /home/max/snap/parity/15594/.local/share/io.parity.ethereum/keys/ethereum
2018-11-13 14:17:42  DB path /home/max/snap/parity/15594/.local/share/io.parity.ethereum/chains/ethereum/db/906a34e69aec8c0d
2018-11-13 14:17:42  State DB configuration: fast
2018-11-13 14:17:42  Operating mode: active
2018-11-13 14:17:42  Configured for Foundation using Ethash engine
2018-11-13 14:17:42  NAT mapped to external address 192.168.1.12:38073
2018-11-13 14:17:42  Public node URL: enode://8b04652d3fc7b2f01f4420d31f945d1cf727360b8d3892e2640c027adaaed0333c8baf09c8525f81ff02cbab70da5402fb42d3011d514ad64deb74afa4720529@192.168.1.12:38073+55037
2018-11-13 14:17:44  Updated conversion rate to Ξ1 = US$208.17 (22875078 wei/gas)
2018-11-13 14:17:52  Syncing snapshot 6/2348        #0    5/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:02  Syncing snapshot 16/2348        #0   13/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:12  Syncing snapshot 29/2348        #0   24/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:22  Syncing snapshot 41/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:32  Syncing snapshot 56/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:42  Syncing snapshot 70/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:52  Syncing snapshot 74/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:02  Syncing snapshot 80/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:07  Syncing snapshot 81/2348        #0   24/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:17  Syncing snapshot 85/2348        #0   26/50 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
```

```
max@max-GF63-8RC:~/.local/share/io.parity.ethereum$ parity --warp-barrier=100000 --no-ancient-blocks
2018-11-13 14:49:15  Starting Parity-Ethereum/v2.1.4-beta-bee2cb8-20181028/x86_64-linux-gnu/rustc1.30.0
2018-11-13 14:49:15  Keys path /home/max/.local/share/io.parity.ethereum/keys/ethereum
2018-11-13 14:49:15  DB path /home/max/.local/share/io.parity.ethereum/chains/ethereum/db/906a34e69aec8c0d
2018-11-13 14:49:15  State DB configuration: fast
2018-11-13 14:49:15  Operating mode: active
2018-11-13 14:49:15  Configured for Foundation using Ethash engine
2018-11-13 14:49:15  NAT mapped to external address 192.168.1.12:46401
2018-11-13 14:49:15  Public node URL: enode://3898da450b882971c52df2d0bf2562d5782dead7520077415576a4359776423a29eba82e94467fd2c1f598be4ded6a01f727bc5882d0d70002f2db413db49351@192.168.1.12:46401+44338
2018-11-13 14:49:17  Updated conversion rate to Ξ1 = US$207.09 (22994372 wei/gas)
2018-11-13 14:49:25  Syncing snapshot 0/2346        #0    6/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:49:30  Syncing snapshot 5/2346        #0    7/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:49:40  Syncing snapshot 10/2346        #0   15/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:49:50  Syncing snapshot 15/2346        #0   18/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:00  Syncing snapshot 18/2346        #0   22/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:10  Syncing snapshot 22/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:20  Syncing snapshot 25/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:25  Syncing snapshot 27/2346        #0   26/50 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:35  Syncing snapshot 32/2346        #0   26/50 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:45  Syncing snapshot 37/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:50:55  Syncing snapshot 42/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:05  Syncing snapshot 46/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:15  Syncing snapshot 48/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:25  Syncing snapshot 54/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:30  Syncing snapshot 57/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:40  Syncing snapshot 63/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:51:50  Syncing snapshot 67/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:52:00  Syncing snapshot 69/2346        #0   24/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:52:05  Syncing snapshot 69/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:52:15  Syncing snapshot 71/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:52:25  Syncing snapshot 74/2346        #0   25/25 peers   8 KiB chain 3 MiB db 0 bytes queue 11 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
```

**Expected Behavior**: Parity should respect `--warp-barrier=100000` and only sync 20 (100000/5000) snapshots and then continue with regular sync.

---
**Previous post**:
I would like to see a configuration option which will allow users to warp sync to first X checkpoints/blocks only and then do a regular sync. 

Motivation/Use case
This will allow users to get all the data they need on a new node as fast as possible. For example, If I wanted event logs from a contract that was created last month then I currently have to wait for full sync to complete which takes days. `warp-upto` option will allow me to warp sync upto a checkpoint just before my contract creation and then regular full sync from thereon. As a result, I will have all the data I need within hours.

`warp-upto` can also work in conjunction with `no-ancient-blocks` to allow users to sync blockchain data from a specific block and not fetch data of previous blocks.

**Comment by maxsam4** [comment]
@joshua-mir Thanks for your quick response. 
I tried `--warp-barrier` before creating the feature request but it did not seem to have worked. I used `parity --warp-barrier=10000` but it still tried to wrap sync all the checkpoints. 
As per the documentation of `--warp-barrier`: "When warp enabled never attempt regular sync before warping to block NUM." It seems like it does the reverse of what I actually want. 
From the documentation, I am inferring that it stops background full sync till warp sync is complete while without `--warp-barrier`, Parity starts a background full sync from block 1 while doing the warp sync.

Please correct me if I am wrong, Thanks.

**Comment by joshua-mir** [comment]
Like you were saying, you need warp-barrier+no-ancient-blocks to do what you want. Presently, ancient block sync (if enabled) happens in the background anyways, but you still have a usable client while it's ongoing.

Perhaps what you want is just the "last checkpoint" from warp sync, if I'm understanding you?

**Comment by maxsam4** [comment]
@joshua-mir warp-barrier + no-ancient-block is not doing what I actually want. Sorry If I wasn't clear earlier. Let me try to explain again.

Currently, there are 2348 checkpoints and if I am not wrong, the default behavior of parity is to warp sync all the 2348 checkpoints and then do a normal sync for the remaining X blocks. 

What I want parity to be able to do is to warp sync upto checkpoint 2000 ONLY and then do a regular sync for remaining 5000*348 + X blocks. I do not want parity to warp sync the latest 5000*348 + X blocks as I want to be able to fetch logs from those blocks ASAP. 

Once parity has completed warp sync to 2000 checkpoint and then regular sync to latest block, it can start regular sync from block 1 depending on if no-ancient-block is specified.

I hope I am making myself clearer.

**Comment by joshua-mir** [comment]
How is warp-barrier behaving differently from that? That is precisely what warp barrier does, (or is supposed to,) in my understanding (except you specify a block number to warp upto and not a checkpoint number...)

**Comment by maxsam4** [comment]
```max@max-GF63-8RC:~/snap/parity/15594/.local/share/io.parity.ethereum$ rm -rf *
max@max-GF63-8RC:~/snap/parity/15594/.local/share/io.parity.ethereum$ parity --warp-barrier=100000 --no-ancient-blocks
2018-11-13 14:17:42  Starting Parity-Ethereum/v2.0.6-unstable-549e202-20180919/x86_64-linux-gnu/rustc1.29.0
2018-11-13 14:17:42  Keys path /home/max/snap/parity/15594/.local/share/io.parity.ethereum/keys/ethereum
2018-11-13 14:17:42  DB path /home/max/snap/parity/15594/.local/share/io.parity.ethereum/chains/ethereum/db/906a34e69aec8c0d
2018-11-13 14:17:42  State DB configuration: fast
2018-11-13 14:17:42  Operating mode: active
2018-11-13 14:17:42  Configured for Foundation using Ethash engine
2018-11-13 14:17:42  NAT mapped to external address 192.168.1.12:38073
2018-11-13 14:17:42  Public node URL: enode://8b04652d3fc7b2f01f4420d31f945d1cf727360b8d3892e2640c027adaaed0333c8baf09c8525f81ff02cbab70da5402fb42d3011d514ad64deb74afa4720529@192.168.1.12:38073+55037
2018-11-13 14:17:44  Updated conversion rate to Ξ1 = US$208.17 (22875078 wei/gas)
2018-11-13 14:17:52  Syncing snapshot 6/2348        #0    5/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:02  Syncing snapshot 16/2348        #0   13/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:12  Syncing snapshot 29/2348        #0   24/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:22  Syncing snapshot 41/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:32  Syncing snapshot 56/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:42  Syncing snapshot 70/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:18:52  Syncing snapshot 74/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:02  Syncing snapshot 80/2348        #0   25/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:07  Syncing snapshot 81/2348        #0   24/25 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
2018-11-13 14:19:17  Syncing snapshot 85/2348        #0   26/50 peers   9 KiB chain 3 MiB db 0 bytes queue 10 KiB sync  RPC:  0 conn,    0 req/s,    0 µs
```

As per the command, It should have synced only first 20 snapshots (100000/5000) but it seems to be syncing all the snapshots.

**Comment by joshua-mir** [comment]
That's definitely not intended... Can you try a recent stable release to see whether this is still the case @maxsam4?

**Comment by maxsam4** [comment]
This seems to be the latest version on snap store. I'll get a newer binary from somewhere else.

**Comment by tomusdrw** [comment]
@maxsam4 Unfortunatelly that's not how snapshots work (they are not checkpoints in the blockchain), the feature you are requesting is not possible (would be unreliable). Let me elaborate.

After every block in Ethereum there is a particular state produced (contains all the accounts, balances, code, and contract storage). This state is stored inside a Patricia-Merkle trie, and root of that tree is part of the header (`state_root`) so that everyone can be sure they got the same result after running all transactions in the block.

Snapshot that we create contains the latest state + a bunch of blocks before that state. Since it's pretty big currently (around 6-7GB) it's split into chunks and that's the stuff that you download.
Snapshots are not created incrementally, so there is no way to re-use previously created chunks (every 10k blocks we start from scratch). When new snapshot is created the old one is discarded.

`--warp-barrier` is trying to download snapshot from block that is greater than the one specified. IIUC you'd like to warp to a previous block. As the old snapshots are removed that would give you a very little chance of success (since the snapshot would be just unavailable), hence we don't really plan to have a feature like that.

As a workaround you can produce the snapshot on your own via CLI (see `parity snapshot --at=<BN>`) and the restore it manually on the other node.

**Comment by maxsam4** [comment]
@tomusdrw That makes sense. I wrongly assumed snapshots were somehow incremental. 
Thanks for the explanation.

**Comment by ulope** [comment]
Then what is the point of the `--warp-barrier` option? I interpreted it exactly as @maxsam4 did since that seems to be the only reason why you would need it.

**Comment by joshua-mir** [comment]
@ulope - it's intended to prevent you getting older snapshots and taking longer to sync as a result. If you don't provide a warp-barrier, you might end up warping to a year-old snapshot created by a node that is not properly in sync.
